lib/puppet/pops/types/p_object_type.rb

```diff
@@ -333,6 +335,76 @@ def initialize(name_or_i12n_hash, i12n_hash_expression = nil)
     end
   end
 
+  # @api private
+  def new_function(loader)
+    @new_function ||= create_new_function(loader)
+  end
+
+  # @api private
+  def create_new_function(loader)
+    impl_name = Loaders.implementation_registry.module_name_for_type(self)
+    if impl_name.nil?
+      # Use generator to create a default implementation
+      impl_class = RubyGenerator.new.create_class(self)
+      class_name = "Anonymous Ruby class for #{name}"
+    else
+      # Can the mapping be loaded?
+      class_name = impl_name[0]
+      impl_class = ClassLoader.provide(class_name)
+
+      raise Puppet::Error, "Unable to load class #{class_name}" if impl_class.nil?
+      unless impl_class < PuppetObject
+        raise Puppet::Error, "Unable to create an instance of #{name}. #{class_name} does not include module #{PuppetObject.name}"
+      end
+    end
+
+    i12n_t = i12n_type
+    from_hash_type = TypeFactory.callable(i12n_t, 1, 1)
+
+    # Create a types and a names array where optional entries ends up last
+    opt_types = []
+    opt_names = []
+    non_opt_types = []
+    non_opt_names = []
+    i12n_t.elements.each do |se|
+      if se.key_type.is_a?(POptionalType)
+        opt_names << se.name
+        opt_types << se.value_type
+      else
+        non_opt_names << se.name
+        non_opt_types << se.value_type
+      end
+    end
+    param_names = non_opt_names + opt_names
+    param_types = non_opt_types + opt_types
+
+    # Create the callable with a size that reflects the required and optional parameters
+    param_types << non_opt_types.size
+    param_types << param_names.size
+    create_type = TypeFactory.callable(*param_types)
+
+    Puppet::Functions.create_loaded_function(:"new_#{name}", loader) do
+      @impl_class = impl_class
+
+      def self.impl_class
+        @impl_class
+      end
+
+      if impl_class.respond_to?(:from_asserted_hash)
+        dispatcher.add_dispatch(from_hash_type, :from_hash, ['hash'], nil, EMPTY_ARRAY, EMPTY_ARRAY, false)
+        def from_hash(hash)
+          self.class.impl_class.from_asserted_hash(hash)
+        end
+      end
+
+      dispatcher.add_dispatch(create_type, :create, param_names, nil, EMPTY_ARRAY, EMPTY_ARRAY, false)
+      def create(*args)
+        self.class.impl_class.new(*args)
+      end
+    end
+  end
```

Comment:
There is quite a lot going on in this method - I think more comments would help a reader (I understand a bit more because I have to context of this PR). For instance what is the @impl_class, what is asserted_hash method.